Added Support for Processing Multiple Tag Sets in Single Script Execution

darkroastcreative opened the pull request:
## Overview

Adds support for downloading posts from e621 for multiple sets of tags without requiring the user to switch out tags between each run. That is, users can specify multiple tag sets in `tags.txt` now (one line per tag set) and the script will iterate over the list of tag sets, downloading posts associated with each set of tags before finishing its execution; eliminating the need for the user to edit `tags.txt` and re-run `e621.py` for each set of tags they'd like to download posts from.

## Definitions

- Tag Set/List — A distinct set of tags, used to search e621 for posts (e.g., `bear male coffee minedoo looking_at_viewer`). A tag set is effectively what you would type into e621 to search for posts. Multiple tag sets would represent different searches against e621 (e.g., bear male coffee minedoo looking_at_viewer` and `hyena leather_jacket male`).

## Notes

- As part of the development process for this feature, I added some comments to the code to help describe what different blocks of code are doing. I did this primarily to help myself understand what the original script was doing in each step of its process, but thought it might be helpful for the maintenance of the script.
- I refactored how file paths are built to use `os.path.join()` to improve cross-platform support, as the original script's means of building file paths didn't work in macOS. In addition, I added constants to represent the paths for `tags.txt` (`TAGS_TXT_FILE_PATH`) and `files.dat` (`FILES_DAT_FILE_PATH`) so these paths don't have to be built every time they're needed.

darkroastcreative committed:
Added support for processing multiple tag sets at a time.

Added support to allow the user to specify multiple tag sets to download posts from, such that the user can download posts from e621 for multiple searches without having to update tags.txt and re-run the script between each set of tags. The convention for doing this is to specify each distinct set of tags to search against on its own line in tags.txt.

darkroastcreative committed:
Updated README to reflect new support for processing multiple tag sets.

Updated the verbiage in the use instructions provided in README to reflect that e621.py can now download posts from multiple tag sets during a single execution.

FiddyFive commented on e621.py:
```diff
@@ -3,85 +3,133 @@
-            url = "https://static1.e621.net/data/"+md5[0:2]+"/"+md5[2:4]+"/"+md5+"."+jsondata["posts"][i]["file"]["ext"]
+# Iterate over the list of tag lists, requesting media from e621 matching each
+# set of tags.
+for tag_list in tags:
+    while True:
```
The page number does not appear to be reset between each list of tags. For example if the first tag_list has 3 pages of results, the next tag_list will pick up from page 3.

darkroastcreative replied:
I pushed a new commit a few moments ago ([49da9f7](https://github.com/FiddyFive/e621-batch-downloader/pull/1/commits/49da9f75d85792b81d825c0d5a6f5a8ae11a8ce3)) which should correct this issue. Effectively, I added a line between lines 64 and 65 as shown in this block to reset `page` to `1` when starting to process a new tag list. Apologies for that issue (something in my gut told me I should've waited to submit the PR until I got some sleep and re-tested my work in the morning)!

FiddyFive commented on e621.py:
```diff
@@ -3,85 +3,133 @@
+        
+        # Move on to request and process the next page of posts from e621.
+        print("Next page.")
+        page += 1
```
See my comment on lines 64&65 about this variable.

FiddyFive reviewed (changes requested):
Found a small bug on page number tracking between tag lists.

darkroastcreative committed:
Added missing page number reset.

Added a line to reset the page number tracker to 1 when starting to process a new tag list, resolving an issue in which not all pages were processed when running the script against multiple tag lists.

FiddyFive commented:
Good work, merged.
